torch/csrc/autograd/engine.cpp

```diff
@@ -623,39 +653,56 @@ auto Engine::execute(const edge_list& roots,
   // Lock post_callbacks_lock_ before clearing final_callbacks_
   ClearCallbacks _cb_guard(final_callbacks_, post_callbacks_lock_);
 
-  GraphTask graph_task(keep_graph, create_graph);
+  std::shared_ptr<GraphTask> graph_task = std::make_shared<GraphTask>(keep_graph, create_graph);
   // Lock mutex while GraphTask is being set up
-  std::unique_lock<std::mutex> lock(graph_task.mutex_);
+  std::unique_lock<std::mutex> lock(graph_task->mutex_);
 
   // Now compute the dependencies for all executable functions and queue the root
   auto graph_root = std::make_shared<GraphRoot>(roots, inputs);
-  compute_dependencies(graph_root.get(), graph_task);
+  compute_dependencies(graph_root.get(), *graph_task);
   if (!outputs.empty()) {
-    graph_task.init_to_execute(*graph_root, outputs);
+    graph_task->init_to_execute(*graph_root, outputs);
   }
-  ready_queue(at::kCPU).push(FunctionTask(&graph_task, std::move(graph_root), InputBuffer(0)));
+  ready_queue(at::kCPU).push(FunctionTask(graph_task, std::move(graph_root), InputBuffer(0)));
 
   // Not a worker
   if (worker_device == NO_DEVICE) {
     // Wait for all tasks to complete
-    graph_task.not_done_.wait(lock, [&graph_task]{
-      return graph_task.outstanding_tasks_.load() == 0;
+    graph_task->not_done_.wait(lock, [&graph_task]{
+      return graph_task->outstanding_tasks_.load() == 0;
     });
   } else {
     // Get back to work while we wait for our new graph_task to
     // complete!
     // See Note [Reentrant backwards]
-    graph_task.owner_ = worker_device;
-    lock.unlock();
-    thread_main(&graph_task);
+    // graph_task.owner_ = worker_device;
+    // lock.unlock();
+    // thread_main(&graph_task);
+    // If no extra threads remaining, create a new thread for reentrant call
```

Comment:
nit: This is the only place where we launch worker threads for reentrant calls, so this isn't really necessary from a code deduplication perspective, but a very logical helper function would be "put some work in the reentrant worker pool, and ensure there is a thread to take care of it", in case there are other situations where we ever need to call in.

Reply:
(The primary benefit of factoring the code out this way, is that you want management of "making sure there are threads to take care of the work" a private implementation detail of the thread pool, rather than something spattered here. But as I said, right now this is pretty low risk, because there's only one site where you need to handle this logic. So this is really just a nit.)